src/lib.rs

```diff
@@ -1,564 +1,570 @@
+        reader.seek(SeekFrom::Current(first_byte as _))?; // position where we want to start reading
+        if start >= end {
+            return Ok(String::new()); // trivial case, empty return result
         }
-    }
 
-    /// Get all soft blocks (lower case blocks) of a chromosome
-    ///
-    /// * `chr` Name of the chromosome from the 2bit file
-    pub fn full_soft_masked_blocks(&self, chr: &str) -> Result<Vec<Block>, Error> {
-        match self.sequence_record(chr) {
-            Ok(record) => Ok(record.soft_mask_blocks),
-            Err(e) => Err(e),
+        let length = end - start;
+        let mut out = Vec::with_capacity(length);
+        unsafe {
+            out.set_len(length);
         }
-    }
 
-    /// Get soft blocks (lower case blocks) of a region on a chromosome
-    ///
-    /// * `chr` Name of the chromosome from the 2bit file
-    pub fn soft_masked_blocks(
-        &self,
-        chr: &str,
-        start: usize,
-        end: usize,
-    ) -> Result<Vec<Block>, Error> {
-        match self.sequence_record(chr) {
-            Ok(record) => {
-                let mut result = Vec::new();
-                for block in record.soft_mask_blocks {
-                    let block_end = block.start + block.length;
-                    if block_end as usize <= start {
-                        continue;
-                    } else if block.start as usize > end {
-                        break;
-                    }
-                    result.push(block)
+        let last_byte = (end - 1) / 4; // inclusive (!) index, and here we know that end >= 1
+        let skip_start = start % 4; // number of pairs to skip in the first byte
+        let partial_start = skip_start != 0;
+
+        let parse_byte =
+            |buf: &mut [u8], r: &mut ValueReader<_>, skip, take, shift| -> Result<()> {
+                let mut byte = r.byte()? << shift;
+                for v in buf.iter_mut().skip(skip).take(take) {
+                    *v = NUC[((byte & 192) >> 6) as usize];
+                    byte <<= 2;
+                }
+                Ok(())
+            };
+
+        if first_byte == last_byte {
+            parse_byte(&mut out, reader, 0, length, skip_start * 2)?; // special case (single byte)
+        } else {
+            // most common case where the sequence is spread over 2 or more bytes
+            let include_end = ((end - 1) % 4) + 1; // number of pairs to include in the last byte
+            let partial_end = include_end != 4;
+
+            if partial_start {
+                parse_byte(&mut out, reader, 0, 4 - skip_start, skip_start * 2)?;
+            }
+            let n_mid = (last_byte - first_byte + 1)
+                - usize::from(partial_start)
+                - usize::from(partial_end);
+            let mut pos = (4 - skip_start) % 4;
+            for _ in 0..n_mid {
+                let byte = reader.byte()? as usize;
+                unsafe {
```

Comment:
Comment about why this is safe

Reply:
I'll check whether something like `[u8]::chunk_exact()` can be used here without performance degradation. Otherwise I'll add a comment.